core/src/main/java/org/wso2/carbon/kernel/internal/configprovider/ConfigProviderImpl.java

```diff
@@ -0,0 +1,241 @@
+    //This regex is used to identify placeholders
+    private static final String PLACEHOLDER_REGEX;
+    //This is used to match placeholders
+    private static final Pattern PLACEHOLDER_PATTERN;
+
+    private ConfigFileReader configFileReader;
+
+    static {
+        PLACEHOLDER_REGEX = "(.*?)(\\$\\{(" + getPlaceholderString() + "):([^,]+?)((,)(.+?))?\\})(.*?)";
+        PLACEHOLDER_PATTERN = Pattern.compile(PLACEHOLDER_REGEX);
+    }
+
+    /**
+     * Enum to hold the supported placeholder types.
+     */
+    private enum Placeholder {
+        SYS("sys"), ENV("env"), SEC("sec");
+        private String value;
+        Placeholder(String value) {
+            this.value = value;
+        }
+        public String getValue() {
+            return value;
+        }
+    }
+
+    public ConfigProviderImpl(ConfigFileReader configFileReader) {
+        this.configFileReader = configFileReader;
+    }
+
+    @Override
+    public <T> T getConfigurationObject(Class<T> configClass) throws CarbonConfigurationException {
+        //get configuration namespace from the class annotation
+        String namespace = null;
+        if (configClass.isAnnotationPresent(Configuration.class)) {
+            Configuration configuration = configClass.getAnnotation(Configuration.class);
+            if (!Configuration.NULL.equals(configuration.namespace())) {
+                namespace = configuration.namespace();
+            }
+        }
+        // lazy loading deployment.yaml configuration.
+        loadDeploymentConfiguration(configFileReader);
+
+        if (namespace != null && deploymentConfigs.containsKey(namespace)) {
+            String jsonConfigString = deploymentConfigs.get(namespace);
+            String yamlConfigString = ConfigurationUtils.convertJSONToYAML(jsonConfigString);
+            if (logger.isDebugEnabled()) {
+                logger.debug("class name: " + configClass.getSimpleName() + " | new configurations : \n" +
+                        yamlConfigString);
+            }
+            String yamlProcessedString = processPlaceholder(yamlConfigString);
+            yamlProcessedString = org.wso2.carbon.kernel.utils.Utils.substituteVariables(yamlProcessedString);
+            Yaml yaml = new Yaml(new CustomClassLoaderConstructor(configClass, configClass.getClassLoader()));
+            yaml.setBeanAccess(BeanAccess.FIELD);
+            return yaml.loadAs(yamlProcessedString, configClass);
+        } else {
+            if (logger.isDebugEnabled()) {
+                logger.debug("Deployment configuration mapping doesn't exist: " +
+                        "creating configuration instance with default values");
+            }
+            try {
+                return configClass.newInstance();
+            } catch (InstantiationException | IllegalAccessException e) {
+                throw new CarbonConfigurationException("Error while creating configuration Instance : "
```

Comment:
Might need to remove the space before the colon: "Instance :". Is there a reason to make the letter "I" uppercase?